actioncable/app/javascript/action_cable/index.js

```diff
@@ -30,16 +30,18 @@ export function getConfig(name) {
 }
 
 export function createWebSocketURL(url) {
-  const webSocketURL = typeof url === "function" ? url() : url
+  return function () {
```

Comment:
I think we can avoid making this function return another function if we move `createWebSocketURL` into `consumer.js`, and call that inside the `url` getter, like:
```diff
diff --git a/actioncable/app/javascript/action_cable/consumer.js b/actioncable/app/javascript/action_cable/consumer.js
index e8440f39f5..e2e0dea8b5 100644
--- a/actioncable/app/javascript/action_cable/consumer.js
+++ b/actioncable/app/javascript/action_cable/consumer.js
@@ -29,11 +29,15 @@ import Subscriptions from "./subscriptions"
 
 export default class Consumer {
   constructor(url) {
-    this.url = url
+    this._url = url
     this.subscriptions = new Subscriptions(this)
     this.connection = new Connection(this)
   }
 
+  get url() {
+    return createWebSocketURL(this._url)
+  }
+
   send(data) {
     return this.connection.send(data)
   }
@@ -52,3 +56,20 @@ export default class Consumer {
     }
   }
 }
+
+export function createWebSocketURL(url) {
+  if (typeof url === "function") {
+    url = url()
+  }
+
+  if (url && !/^wss?:/i.test(url)) {
+    const a = document.createElement("a")
+    a.href = url
+    // Fix populating Location properties in IE. Otherwise, protocol will be blank.
+    a.href = a.href
+    a.protocol = a.protocol.replace("http", "ws")
+    return a.href
+  } else {
+    return url
+  }
+}
diff --git a/actioncable/app/javascript/action_cable/index.js b/actioncable/app/javascript/action_cable/index.js
index e679745fd7..848b5631d6 100644
--- a/actioncable/app/javascript/action_cable/index.js
+++ b/actioncable/app/javascript/action_cable/index.js
@@ -1,6 +1,6 @@
 import Connection from "./connection"
 import ConnectionMonitor from "./connection_monitor"
-import Consumer from "./consumer"
+import Consumer, { createWebSocketURL } from "./consumer"
 import INTERNAL from "./internal"
 import Subscription from "./subscription"
 import Subscriptions from "./subscriptions"
@@ -15,11 +15,12 @@ export {
   Subscription,
   Subscriptions,
   adapters,
+  createWebSocketURL,
   logger,
 }
 
 export function createConsumer(url = getConfig("url") || INTERNAL.default_mount_path) {
-  return new Consumer(createWebSocketURL(url))
+  return new Consumer(url)
 }
 
 export function getConfig(name) {
@@ -28,18 +29,3 @@ export function getConfig(name) {
     return element.getAttribute("content")
   }
 }
-
-export function createWebSocketURL(url) {
-  const webSocketURL = typeof url === "function" ? url() : url
-
-  if (webSocketURL && !/^wss?:/i.test(webSocketURL)) {
-    const a = document.createElement("a")
-    a.href = webSocketURL
-    // Fix populating Location properties in IE. Otherwise, protocol will be blank.
-    a.href = a.href
-    a.protocol = a.protocol.replace("http", "ws")
-    return a.href
-  } else {
-    return webSocketURL
-  }
-}
```

What do you think?

Reply:
yeah I like that refactor, I was trying to avoid changing too much in the codebase but this doesn't change anything in the public facing API. I like signaling `_url` as private with the underscore too.